### DISCUSSION TITLE
Linkerd in HA with 2 controllerReplicas

### DISCUSSION THREAD

**Discussion by listenerme** [linkerd/linkerd2#4618] [General]
## Bug Report

As per the document https://linkerd.io/2/features/ha/, it looks we can limit the controller replicas to 2 for HA installation. I have deployed the same way in one node , it says "unschedulable pods" for the pods supposed to be scheduled on the second node. Second time, brought the second node up, old error is gone. However,  shutting down one node fails the api, and it throws "error upgrading connection: error dialing backend: dial tcp <NODEIP>:10250: i/o timeout"

### What is the issue?

Linkerd API is inaccessible while running with 1 node with HA configuration(2 replicas)

### How can it be reproduced?
1) Change controllerReplicas in values-ha.yaml to 2
2) Deploy linkerd with (-f values-ha.yaml), but in a single node with podantiaffinity(true by default)
3) Add a second node, it will schedule all pods.. Again shut any of the node, controller api fails right away.

### Logs, error output, etc
stat pods
Cannot find Linkerd: error upgrading connection: error dialing backend: dial tcp NODEIP:10250: i/o timeout
Validate the install with: linkerd check

(If the output is long, please create a [gist](https://gist.github.com/) and
paste the link here.)

#### `linkerd check` output
× can initialize the client
    error upgrading connection: error dialing backend: dial tcp NODEIP:10250: i/o timeout
    see https://linkerd.io/checks/#l5d-existence-client for hints

```text
your output here ...
```

### Environment

- Kubernetes Version: 1.15.11
- Cluster Environment: (GKE, AKS, kops, ...) AKS
- Host OS: ubuntu
- Linkerd version: 2.8.0

**Comment by grampelberg** [comment]
What API fails?

**Comment by listenerme** [comment]
Its not api failure. Even with all required pods are up in one of the node(with second replicas are pending), unable to call the api(with linkerd stat). When checked with "linkerd check", same behavior and it halted at unscheduled pods

Isnt it supposed to be working even with single replicas?

**Reply by grampelberg** [reply]
`stat` should work if everything is up and running (including prometheus and linkerd-controller). You need to run 3 nodes to use `--ha` and have `check` pass.

**Reply by listenerme** [reply]
Does that mean 3 replicas(on 3 nodes) required only for passing "linkerd check"? Linkerd api will be accessible and we can mesh application too with single replica right in case of two node failure?

Also can we just include 3 replicas on 2 nodes too?

**Reply by grampelberg** [reply]
> Does that mean 3 replicas(on 3 nodes) required only for passing "linkerd check"?

In HA mode, yes.

> Linkerd api will be accessible and we can mesh application too with single replica right in case of two node failure?

Yes.

> Also can we just include 3 replicas on 2 nodes too?

Yes, but not in `--ha` mode.

**Reply by listenerme** [reply]
I have just tested this, two different ways, api failed in both cases
1) **Shutting down one node**: api not available
2) **Drained one node**: api not available

See the error (in next post for readability ) when I try to "linkerd stat" or "linkerd inject"

**Comment by listenerme** [comment]
]$ **linkerd -n linkerd stat deploy**
Cannot find Linkerd: linkerd-controller-64cf8696f7-mf869: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-destination-57847f975b-fxlj6: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-identity-5bdd667895-jrz2j: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-proxy-injector-5689db6474-jpdt6: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-sp-validator-84d86bf697-2r2vx: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-tap-799498f9f8-c8lqq: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
Validate the install with: linkerd check


$ **kubectl get deploy siz -o yaml | linkerd inject - | kubectl apply -f -**
Cannot find Linkerd: linkerd-controller-64cf8696f7-mf869: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-destination-57847f975b-fxlj6: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-identity-5bdd667895-jrz2j: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-proxy-injector-5689db6474-jpdt6: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-sp-validator-84d86bf697-2r2vx: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
    linkerd-tap-799498f9f8-c8lqq: 0/2 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing pods anti-affinity rules, 1 node(s) were unschedulable.
Validate the install with: linkerd check
error: no objects passed to apply

**You can see all necessary pods are up in one node, other node, its pending.**
linkerd-controller-64cf8696f7-mf869        0/2     Pending   0          40s
linkerd-controller-64cf8696f7-pzkgx        2/2     Running   1          34m
linkerd-destination-57847f975b-fxlj6       0/2     Pending   0          40s
linkerd-destination-57847f975b-hw89x       2/2     Running   1          34m
linkerd-grafana-75c8d56f96-sg8v6           2/2     Running   0          34m
linkerd-identity-5bdd667895-5hzjq          2/2     Running   0          34m
linkerd-identity-5bdd667895-jrz2j          0/2     Pending   0          40s
linkerd-prometheus-6d8f85b995-77h5l        2/2     Running   0          14m
linkerd-proxy-injector-5689db6474-fwp4r    2/2     Running   0          34m
linkerd-proxy-injector-5689db6474-jpdt6    0/2     Pending   0          40s
linkerd-sp-validator-84d86bf697-2r2vx      0/2     Pending   0          40s
linkerd-sp-validator-84d86bf697-pv975      2/2     Running   0          34m
linkerd-tap-799498f9f8-8srh6               2/2     Running   1          34m
linkerd-tap-799498f9f8-c8lqq               0/2     Pending   0          40s
linkerd-web-c9cb6bdb6-dk5vz                2/2     Running   1          34m
